x-pack/plugins/ingest_manager/server/routes/datasource/handlers.ts

```diff
@@ -76,23 +77,43 @@ export const createDatasourceHandler: RequestHandler<
   const soClient = context.core.savedObjects.client;
   const callCluster = context.core.elasticsearch.legacy.client.callAsCurrentUser;
   const user = (await appContextService.getSecurity()?.authc.getCurrentUser(request)) || undefined;
-  const newData = { ...request.body };
+  const logger = appContextService.getLogger();
+  let newData = { ...request.body };
   try {
+    // If we have external callbacks, then process those now before creating the actual datasource
+    const externalCallbacks = appContextService.getExternalCallbacks('datasourceCreate');
+    if (externalCallbacks && externalCallbacks.size > 0) {
+      let updatedNewData: NewDatasource = newData;
+
+      for (const callback of externalCallbacks) {
+        try {
+          updatedNewData = await callback(updatedNewData);
+        } catch (error) {
+          // Log the error, but keep going and process the other callbacks
+          logger.error('An external registered [datasourceCreate] callback failed when executed');
+          logger.error(error);
+        }
+      }
+
+      // @ts-ignore
```

Comment:
I needed this ignore because because for some reason I was not able to just re-assign the `updatedNewData` back to `newData`, so I had to cast it below and even that was not working. I think its because the type used here (from `CreateDatasourceRequestSchema.body` is created via `schema` - maybe?)

Reply:
I'd like to find the root cause for this and fix it / remove the `ts-ignore`.

At a minimum I'd like the reason for the ignore to be put next to it so we know why it's there for the next person who sees & wants to remove

Reply:
I'll try to spend some more time playing with it to see to see if I can figure it out 👍